### PR TITLE
Fix build on OpenBSD

### PR DESCRIPTION
**Pull request by hubslave**
Refer to the commit messages for details.

The first (fix swap functions on Bitrig and OpenBSD) and last (use std::experimental::string_view in json) changes are particularly hacky, but unless the project is willing to rename the entire Common::swap* function family, I can't think of a better way to do the former. The latter can just be dropped if it's viewed as too hacky; it's understandable if the project doesn't want to cater to incomplete c++17 stdlib implementations.

With these changes, citra compiles on OpenBSD with the following commands:
```
mkdir build && cd build
export Qt5_DIR=/usr/local/lib/qt5/cmake/Qt5
cmake \
    -DCMAKE_CXX_FLAGS='-I/usr/local/include -O2' \
    -DCMAKE_EXE_LINKER_FLAGS=-zwxneeded \
    -DUSE_SYSTEM_CURL=1 ..
make
```

As an aside, I noticed that compiling with `-DCMAKE_BUILD_TYPE=Release` produces a very slow build on OpenBSD; is this a symptom of the default compiler being Clang? I have to compile with `-O` manually set to >=2 to get playable games.

---
This change is [Reviewable](https://reviewable.io/reviews/citra-emu/citra/3456)

### CODE REVIEW

File: src/core/hle/service/service.cpp [hunk 1]
```diff
@@ -150,15 +150,16 @@ void ServiceFrameworkBase::ReportUnimplementedFunction(u32* cmd_buf, const Funct
...
     for (int i = 1; i <= num_params; ++i) {
-        w.write(", [{}]={:#x}", i, cmd_buf[i]);
+        fmt::format_to(buf, ", [{}]={:#x}", i, cmd_buf[i]);
     }
-    w << '}';
+    buf.push_back('}');
 
-    LOG_ERROR(Service, "unknown / unimplemented %s", w.c_str());
+    buf.push_back('\0');
+    LOG_ERROR(Service, "unknown / unimplemented %s", &buf[0]);
```
Comment (TTFH) [inline src/core/hle/service/service.cpp]: fmt::to_string(buf)

**Comment by lioncash** [comment]
Personally I don't agree with applying modifications to external libs. It introduces bookkeeping, and potentially (and can potentially make updating libraries a nuisance; not in this case, of course, but still, we shouldn't allow it). It would be better to submit the change to the original repo, which we then pull downstream.

File: src/core/hle/service/soc_u.cpp [hunk 2]
```diff
@@ -94,7 +94,9 @@ static const std::unordered_map<int, int> error_map = {{
     {ERRNO(EMFILE), 33},
     {EMLINK, 34},
     {ERRNO(EMSGSIZE), 35},
+#ifdef EMULTIHOP
```
Comment (lioncash) [inline src/core/hle/service/soc_u.cpp]: This should have a comment by it explaining its purpose.

Comment (hubslave) [reply src/core/hle/service/soc_u.cpp]: The other #ifdefs in the array, introduced for a similar purpose in commit 702439b, don't have comments; won't it be confusing to have one odd man out? Or do you mean to comment all of them, and just start with this?

**Comment by hubslave** [comment]
> It would be better to submit the change to the original repo, which we then pull downstream.

This particular change was already closed as wontfix in nlohmann/json#735; upstream's advice is to use workarounds on affected systems. (The exact fix proposed in the thread, of just including `<experimental/string_view>` before `<json.hpp>`, didn't work for me, because `json.hpp` still refers to `std::string_view`, not `std::experimental::string_view`.)

**Comment by wwylele** [comment]
> This particular change was already closed as wontfix in nlohmann/json#735; upstream's advice is to use workarounds on affected systems. (The exact fix proposed in the thread, of just including <experimental/string_view> before <json.hpp>, didn't work for me, because json.hpp still refers to std::string_view, not std::experimental::string_view.)

And I agree with and want to respect to the upstream decision. (So even you want to "fix" it, leave it outside the external directory). Also, using experimental namespace would break those platform where string_view is already moved out from the experimental.

File: src/core/hle/service/service.cpp (hunk 1 above)
Comment (hubslave) [reply src/core/hle/service/service.cpp]: LOG_ERROR() expects a C string; I could replace `&buf[0]` with `fmt::to_string(buf).c_str()`, but whether or not that is safe depends on whether LOG_ERROR() is always implemented as a full expression. Is the project willing to rely on that?

If not, one would have to assing `fmt::to_string(buf)` to a temporary, but that would result in a pointless copy; if the concern is code clarity, I think a comment would do just as well.

**Comment by hubslave** [comment]
OK; I updated the pull request with an alternative patch that introduces a wrapper header `src/web_service/json.h` that substitutes `experimental::string_view` in for `string_view` if the latter is not available while the former is for the duration of `#include <json.hpp>`. This did require patching all consumers to include the new `json.h` instead of `json.hpp`, but thankfully there were only three (all under `src/web_service`), and `json.hpp` is now free of changes.

File: src/web_service/json.h [hunk 3]
```diff
@@ -0,0 +1,14 @@
+// This file is under the public domain.
```
Comment (B3n30) [inline src/web_service/json.h]: Citra header?

File: src/core/hle/service/service.cpp (hunk 1 above)
Comment (B3n30) [reply src/core/hle/service/service.cpp]: I think fmt::to_string(buf).c_str() is fine

**Comment by hubslave** [comment]
Pull request updated to address review comments.

I also changed the externals/fmt patch to use an even newer version of fmtlib to see if it helps with the msvc build error [1].

How seriously should the clang-format build failure [2] be taken? I notice in particular that it asks to change `__has_include(<experimental/string_view>)` to `__has_include(<experimental / string_view>)`, which of course won't work. It also complains about a bit of indentation in src/core/hle/service/service.cpp, but I didn't touch that line; does clang-format just have different indentation rules for member vs. free function calls?

[1]: https://ci.appveyor.com/project/bunnei/citra/build/1.0.7360/job/bhwvs4r3pk0irk18
[2]: https://travis-ci.org/citra-emu/citra/jobs/345728265

**Comment by wwylele** [comment]
@hubslave For now you can ignore clang-format error as it is outdated and an update is incoming

**Comment by jroweboy** [comment]
>  I notice in particular that it asks to change __has_include(<experimental/string_view>) to __has_include(<experimental / string_view>), 

this indicates you should probably wrap that section in a clang format off block so it won't mess with it

**Comment by hubslave** [comment]
OK; I also fixed the src/core/hle/service/service.cpp line to match what clang-format demanded since it was such a little thing. Hopefully this build is all green...

**Pull request merged**